### PR TITLE
fix: use viper to get configuration strings

### PR DESCRIPTION
**Pull request by remyleone**
I would like to use a configuration file stored in `~/.kops.yaml` to store my credentials and many configuration preferences that will remain constant across several cluster creations. Therefore, I think it would be useful to use the helper `viper.GetString` to get information from the configuration file and in case they are not defined there, it will be go to the environment variables.

I guess we would need to do that across all Kops. At the moment many configuration options are not working for me when they are defined in the configuration file and not with environment variables without the use of `viper.GetString()`.

### CODE REVIEW

**Comment by hakman** [comment]
Running `make gomod` should fix the missing deps.

**Comment by remyleone** [comment]
@hakman I'm not sure about how to mock credentials in the context of this test

**Comment by remyleone** [comment]
I've tried to do a search and replace across the whole code base and it seems to work except for this test

**Comment by johngmyers** [comment]
Viper isn't seeing the change to the environment variable that is done in the test setup. The test setup probably has to be changed to push the value into viper (and clear it aterwards).

File: util/pkg/vfs/tests/s3fs_test.go [hunk 1]
```diff
@@ -51,9 +53,8 @@ func TestS3RenderTerraform(t *testing.T) {
 			`,
 		},
 	}
-	t.Setenv("S3_ENDPOINT", "foo.s3.amazonaws.com")
-
-	t.Setenv("KOPS_STATE_S3_ACL", "bucket-owner-full-control")
+	viper.Set("S3_ENDPOINT", "foo.s3.amazonaws.com")
+	viper.Set("KOPS_STATE_S3_ACL", "bucket-owner-full-control")
```
Comment (hakman) [inline util/pkg/vfs/tests/s3fs_test.go:57]: This is not really required. The problem is mainly that `viper.BindEnv("ENV_VAR")` has to be called before using `viper.GetString("ENV_VAR")`.

File: util/pkg/vfs/s3context.go [hunk 2]
```diff
@@ -146,10 +148,10 @@ func (s *S3Context) getDetailsForBucket(bucket string) (*S3BucketDetails, error)
 	}
 
 	// Probe to find correct region for bucket
-	endpoint := os.Getenv("S3_ENDPOINT")
+	endpoint := viper.GetString("S3_ENDPOINT")
```
Comment (hakman) [inline util/pkg/vfs/s3context.go:151]: Calling `viper.BindEnv()` should make env reading possible. A little annoying, but not sure how to go about it differently at the moment.
```suggestion
	_ = viper.BindEnv("S3_ENDPOINT")
	endpoint := viper.GetString("S3_ENDPOINT")
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by remyleone** [comment]
@hakman could you review again? :)

File: pkg/featureflag/featureflag.go [hunk 3]
```diff
@@ -39,7 +40,7 @@ const (
 )
 
 func init() {
-	ParseFlags(os.Getenv(Name))
+	ParseFlags(viper.GetString(Name))
```
Comment (hakman) [inline pkg/featureflag/featureflag.go:43]: I think `viper.BindEnv()` would be required here also.
```suggestion
	_ = viper.BindEnv(Name)
	ParseFlags(viper.GetString(Name))
```

**Review by hakman — CHANGES_REQUESTED** [review]

**Comment by hakman** [comment]
@johngmyers Could you take a look too?

**Comment by johngmyers** [comment]
I think we need a coherent strategy for using viper, probably that of https://github.com/carolynvs/stingoftheviper. I would prefer the viper keys be lowercase hyphen-separated, using the flag names, not all-caps underscore-separated.

**Comment by hakman** [comment]
@remyleone Would you mind rebasing to address https://github.com/kubernetes/kops/pull/14642#issuecomment-1462082897?

**Comment by hakman** [comment]
@remyleone Could you remove the go.mod & go.sum changes and squash the commits?

**Comment by johngmyers** [comment]
I think my comments of Dec 18, 2022 are still relevant. We definitely shouldn't be defining new all-caps viper keys.

**Comment by hakman** [comment]
> I think my comments of Dec 18, 2022 are still relevant. We definitely shouldn't be defining new all-caps viper keys.

To answer your comment from Dec 18, https://github.com/carolynvs/stingoftheviper may work in the example, but could not get it to work in practice.

Also, I only want to redo the way we do things for the existing env vars for cloud provider, cluster, state store and some other things used for development.

**Pull request closed without merging**